### ISSUE TITLE
Expose Openshift Docker Registry

### ISSUE THREAD

**Issue by glennodickson** [minishift/minishift#254]
Environment:
minishift 0.9.0
OS : Windows 10
Docker client : 1.9.1

I'm trying to login/push into Openshift Integrated Docker Registry. Below I have describe both of these and their errors. **If anyone can help me through this I would be most appreciative.**

# Started Minishift
I've started minishift (not is used correctly) from the window command prompt using:
`minishift start --insecure-registry 172.30.0.0/16    --insecure-registry minishift 
`

# Registry Service 
```
c:\..\..\>oc describe service docker-registry
Name:                   docker-registry
Namespace:              default
Labels:                 docker-registry=default
Selector:               docker-registry=default
Type:                   NodePort
IP:                     172.30.225.22
Port:                   5000-tcp        5000/TCP
NodePort:               5000-tcp        30832/TCP
Endpoints:              172.17.0.2:5000
Session Affinity:       ClientIP
No events.
```

# Registry Route:
```
apiVersion: v1
kind: Route
metadata:
  name: docker-registry
  namespace: default
  selfLink: /oapi/v1/namespaces/default/routes/docker-registry
  uid: 9f7eee66-d299-11e6-89fb-42527179d16d
  resourceVersion: '88756'
  creationTimestamp: '2017-01-04T16:19:52Z'
  labels:
    docker-registry: default
  annotations:
    openshift.io/host.generated: 'true'
spec:
  host: docker-registry-default.192.168.99.100.xip.io
  to:
    kind: Service
    name: docker-registry
    weight: 100
  port:
    targetPort: 5000-tcp
  tls:
    termination: edge
    insecureEdgeTerminationPolicy: Allow
status:
  ingress:
    -
      host: docker-registry-default.192.168.99.100.xip.io
      routerName: router
      conditions:
        -
          type: Admitted
          status: 'True'
          lastTransitionTime: '2017-01-04T16:19:52Z'
```
# Login
Attempt to login from the window command prompt using:
 ` docker login  -u admin -p admin -e email@email.com docker-registry-default.192.168.99.100.xip.io`

Get this error:
 
> Error response from daemon: Get **https**://docker-registry-default.192.168.99.100.xip.io/v1/users/: x509: certificate is valid for router.default.svc, router.default.svc.cluster.local, not docker-registry-default.192.168.99.100.xip.io

# Push
List Docker images from WIndows command prompt:
```
c::\....\..>docker images
REPOSITORY                                                                       TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
docker-registry-default.192.168.99.100.xip.io:30832/fabric8/camel-rest-service   1.0-SNAPSHOT        sha256:7e1bb        26 hours ago        572 MB
```

Attempt to Push from the window command prompt using:
`docker push docker-registry-default.192.168.99.100.xip.io:30832/fabric8/camel-rest-service:1.0-SNAPSHOT

Get error:

> The push refers to a repository [docker-registry-default.192.168.99.100.xip.io:30832/fabric8/camel-rest-service]
> Get https://docker-registry-default.192.168.99.100.xip.io:30832/v1/_ping: tls: oversized record received with length 20527
> 

Two issues here :
1. Certificate issue - not valid for docker-registry-default.192.168.99.100.xip.io:30832
2. Https being used to access Openshift Integrated Docker Registry @ docker-registry-default.192.168.99.100.xip.io:30832

Any pointers as how to solve would be great,

**Comment by LalatenduMohanty** [comment]
@glennodickson I have not tried myself but looks like these steps worked for 1.0.0-beta-1 https://github.com/minishift/minishift/issues/220
 I would suggest you to update to 1.0.0-beta.1 if possible.

**Comment by glennodickson** [comment]
Thanks for the response. With #220 he alters his route's host name to match that of the certificate to:
`   host: 192.168.64.2.xip.io`

I tried that but it didn't work. I'll delete my minishift cluster and start again to see if that makes a difference and retry it.

I would have thought regenerating the certificate and maybe using wildcard names might be better? I'm very new to Openshift so I might be getting it wrong. Bit like this post 
https://github.com/kubernetes/heapster/issues/571

**Comment by glennodickson** [comment]
I've reinstalled minishift and tried to connect to the Openshift Registry.

Started Minishift as suggested by #220  using the --insecure-registry switch:
`minishift start --insecure-registry 172.30.0.0/16  --insecure-registry minishift`

Added the route that points to Openshift Registry:

Description of Pod, Service, route

**C:\...>oc get pods**
```
NAME                      READY     STATUS    RESTARTS   AGE
docker-registry-1-tq6et   1/1       Running   4          8h
```

**oc get services**
```
NAME              CLUSTER-IP     EXTERNAL-IP   PORT(S)                 AGE
docker-registry   172.30.19.60   <nodes>       5000/TCP                9h
kubernetes        172.30.0.1     <none>        443/TCP,53/UDP,53/TCP   9h

```
**oc get route**
```
NAME              HOST/PORT                                       PATH      SERVICES          PORT       TERMINATION
docker-registry   docker-registry-default.192.168.99.100.xip.io             docker-registry   5000-tcp   edge/Allow
```

**oc describe route docker-registry**
```
Name:                   docker-registry
Namespace:              default
Created:                7 hours ago
Labels:                 docker-registry=default
Annotations:            openshift.io/host.generated=true
Requested Host:         docker-registry-default.192.168.99.100.xip.io
Path:                   <none>
TLS Termination:        edge
Insecure Policy:        Allow
Endpoint Port:          5000-tcp

Service:        docker-registry
Weight:         100 (100%)
Endpoints:      172.17.0.2:5000
```

**Login using Docker with host docker-registry-default.192.168.99.100.xip.io:8443**

```
C:\...>docker login -u admin -p 7zxJrIji1fqfn1wp93rZHkGK4hkkahCg7KjId4EBe68 docker-registry-default.192.168.99.100.xip.io:8443
Email:
Error response from daemon: Get https://docker-registry-default.192.168.99.100.xip.io:8443/v1/users/: x509: certificate is valid for kubernetes, kubernetes.default, kubernetes.default.svc, kubernetes.default.svc.cluster.local, localhost, openshift, openshift.default, openshift.default.svc, openshift.default.svc.cluster.local, 10.0.2.15, 127.0.0.1, 172.17.0.1, 172.30.0.1, 192.168.99.100, not docker-registry-default.192.168.99.100.xip.io
```
**Login using Docker with host 192.168.99.100:8443**
```
C:\...>docker login -u admin -p 7zxJrIji1fqfn1wp93rZHkGK4hkkahCg7KjId4EBe68 192.168.99.100:8443
Email:
Error response from daemon: Get https://192.168.99.100:8443/v1/users/: x509: certificate signed by unknown authority
```
It's unclear what the host should be eventhough I've followed the #220 advice to use the --insecure-registry during the Minishift start up. 

Being new Openshift and Docker I'm really unsure, any comments/advice would welcome.

**Comment by praveenkumar** [comment]
@glennodickson Can you please check https://github.com/minishift/minishift/issues/277#issuecomment-272689327 steps and try out.

**Comment by hferentschik** [comment]
@LalatenduMohanty have you seen https://github.com/minishift/minishift/issues/177#issuecomment-279218065? Whatever we do here, should probably somehow be a special case of whatever we are going to to for #177.

In the easiest version we would just introduce the concept of a _Package_ and _OpenShiftConfigurator_ to have these basic pieces in place. Moving forward we would then expose these concepts in form of some commands as part of #177. 

Or what were your plans?

**Comment by chiippy** [comment]
As Vagrant openshift isn’t working  under windows now, so I have had a look at minishift (1.0.0-beta.4) and getting this problem when open up the docker-registry. Seems like no a resolution at the moment.

Looking in the dockerd, the docker-registry service URL isn’t marked as insecure.  I can change the docker.service to include this but surely no a simple way.

So, I performed :
1)	Minishift delete, then  record the allocated IP.
2)	Starts the new image :
minishift.exe  start --vm-driver="virtualbox"  --memory 4096 --cpus 3 --insecure-registry 172.30.0.0/16 --insecure-registry minishift  --insecure-registry docker-registry-default.**192.168.99.103**.xip.io

adding the new docker-registry URL in the startup.

After docker login, docker push works nicely :

C:\minis>docker push  docker-registry-default.192.168.99.103.xip.io/fxdev/amq
The push refers to a repository [docker-registry-default.192.168.99.103.xip.io/fxdev/amq]
9d72141b4dff: Pushed
......
latest: digest: sha256:1e25b358884200eb9d2cb265530df3c6cc9489b43591ead0f78de8373d3e69a8 size: 19537

C:\minis>oc get is
NAME      DOCKER REPO                   TAGS      UPDATED
amq       172.30.20.75:5000/fxdev/amq   latest    3 minutes ago

No a proper solution but a simpler walk around to get me going.  

I guess if we can have an optional flag at startup to indicate to the minishift that the docker-registry should be marked as insecure-registry ? So that I don't need to second guess the new IP address will be a better alternative ? 

Regards

Alex

**Comment by budhrg** [comment]
@chiippy , thanks for the workaround. 

The last thing we tried is mentioned here https://github.com/minishift/minishift/issues/277#issuecomment-272689327  which seems to be not working from host.

**Comment by chiippy** [comment]
I am using Windows 7 + VirtualBox. minishift uses the boot2docker.iso image.  Looking at the comments in #277. I am a bit lost about 'no working from host ?'.  I guess host mean the VirtualBox VM ?  I am running minishift from my PC and I didn't change  anything inside the VM, so I don't have the problem losing changes inside the VM after minishift stop/start.  
As you can see from my previous reply, I am simply using the DOS prompt (no powershell) to issue my docker or oc commands.  All seems fine, no much different from if I deploy my stuffs into AWS.

**Comment by LalatenduMohanty** [comment]
@chiippy Some hypervisors e.g. Xhyve allocates new IP every time we do `minishift start` , so your suggested method is not applicable to all the hypervisors Minishift supports.

**Comment by LalatenduMohanty** [comment]
Update: While trying the manual steps on CentOS ISO everything works fine except I am seeing an issue i.e. error `unauthorized: authentication required` while pushing the docker image to the local docker registry. I am looking in to it and will keep it posted here. 

##Stage -1 (On the Host) 
```
$ minishift start --iso-url file:///home/lmohanty/Downloads/minishift-centos7.iso

$ export PATH=$PATH:~/.minishift/cache/oc/v1.4.1

$ oc login -u system:admin

$ oc adm policy add-cluster-role-to-user cluster-admin developer

$ oc expose service docker-registry -n default
 
$ oc get routes -n default
NAME              HOST/PORT                                       PATH      SERVICES          PORT       TERMINATION
docker-registry   docker-registry-default.192.168.42.83.nip.io             docker-registry   5000-tcp   

```
##Stage -2 (In the CentOS VM)

```
$ minishift ssh

# Add insecure registry "docker-registry-default.192.168.64.2.xip.io:80" to docker service file in /etc/systemd/system/docker.service


$ sudo systemctl daemon-reload
$ docker stop origin
$ sudo systemctl restart docker
$ docker start origin
```
`minishift stop and then start` will not work as /etc/systemd/system/docker.service will be over written for each minishift start.

##Stage-3 (On the host)
```
$ oc whoami
system:admin
$ oc logout

$ oc login -u developer -p developer

$ oc whoami -t
qWtFnC3Dj9aizZp5JTURp5YO9frBALeRgWlBGKKTs5A

$ eval $(minishift docker-env)

$ docker login -u developer -p 5lsWLLaKeIpXUGtGnPMzK7R0aUDojW8o91wfLojqrGw docker-registry-default.192.168.42.83.nip.io:80
Login Succeeded

```
##Stage-4 (On the Host) :  Push a docker image to the local registry

```
$ docker pull busybox

$ docker tag docker.io/busybox docker-registry-default.192.168.42.83.nip.io:80/myproject/busybox

$ docker push docker-registry-default.192.168.42.83.nip.io:80/openshift/busybox

The push refers to a repository [docker-registry-default.192.168.42.83.nip.io:80/openshift/busybox]
38ac8d0f5bb3: Pushing [==================================================>] 1.143 MB
unauthorized: authentication required

```

**Comment by jorgemoralespou** [comment]
@hferentschik @praveenkumar @LalatenduMohanty @budhrg @gbraad The VMs need to provide the same IP for every minishift restart after it has been created. It don't necessary need to be the same between minishift instantiations, but it would also be nice. Like with Vagrant, were you provided a fixed IP.

There's currently two problems, on is that oc cluster will not start up if the IP you get between restarts is not the same, and the second is that it will be very difficult to expose the registry.

For the later, you can probably do things like port-forward and configure localhost:5000 to expose the registry locally, but this is a hack.

**Comment by gbraad** [comment]
@jorgemoralespou agree, we might have to bump #343

**Comment by LalatenduMohanty** [comment]
@jorgemoralespou At this point we are dependent on hypervisors to get the IP address. I think we can put some check between start->stop->start to see if we are getting the same IP address else we can give feedback to the user that we need to run delete and then start. WDYT?

**Comment by LalatenduMohanty** [comment]
I have added few more privileges (as below) as per [1]  to the user  in *Stage-1* i.e. https://github.com/minishift/minishift/issues/254#issuecomment-284985284 , but I am still seeing the issue. 
```
$ minishift start --iso-url file:///home/lmohanty/Downloads/minishift-centos7.iso

$ export PATH=$PATH:~/.minishift/cache/oc/v1.4.1

$ oc login -u system:admin

$ oc adm policy add-cluster-role-to-user cluster-admin developer

$ oc adm policy add-role-to-user system:registry developer
$ oc adm policy add-role-to-user admin developer -n openshift
$ oc adm policy add-role-to-user system:image-builder developer

$ oc expose service docker-registry -n default
 
$ oc get routes -n default
NAME              HOST/PORT                                       PATH      SERVICES          PORT       TERMINATION
docker-registry   docker-registry-default.192.168.42.83.nip.io             docker-registry   5000-tcp   

```
[1] https://docs.openshift.org/latest/install_config/registry/accessing_registry.html

**Comment by LalatenduMohanty** [comment]
Sent a mail to OpenShift development list and asked for suggestions http://lists.openshift.redhat.com/openshift-archives/dev/2017-March/msg00005.html

**Comment by hferentschik** [comment]
> oc adm policy add-cluster-role-to-user cluster-admin developer
> oc adm policy add-role-to-user admin developer -n openshift

Really? Why is that? Now we are opening up everything. Why are these needed. Are you not effectively making the developer an admin?

**Comment by praveenkumar** [comment]
> Are you not effectively making the developer an admin?

That only for some testing purpose I think when things were not working as expected.

**Comment by hferentschik** [comment]
> That only for some testing purpose I think when things were not working as expected.

Fair enough

**Comment by LalatenduMohanty** [comment]
@hferentschik Yup, that was for testing purpose.

**Comment by LalatenduMohanty** [comment]
I could reproduce the issue on an "oc cluster up" setup on Fedora 25. 
```
$ docker push docker-registry-default.10.70.7.55.xip.io:80/myproject/busybox 
The push refers to a repository [docker-registry-default.10.70.7.55.xip.io:80/myproject/busybox]
c0de73ac9968: Pushing [==================================================>] 1.143 MB
unauthorized: authentication required
```

**Comment by hferentschik** [comment]
@thatdocslady and that's after you successfully logged into the registry?

**Comment by LalatenduMohanty** [comment]
@hferentschik yes, that is right. I have updated the previous comment.

**Comment by LalatenduMohanty** [comment]
As discussed with @miminar it is a duplicate of https://bugzilla.redhat.com/show_bug.cgi?id=1383439 . It is fixed in 1.5.0. 

I tested  local docker registry (pull/push) successfully through "oc cluster up" in 1.5.0-rc.0  [1] 

```
$ ./oc login -u system:admin
Logged into "https://192.168.2.5:8443" as "system:admin" using existing credentials.

You have access to the following projects and can switch between them with 'oc project <projectname>':

    default
    kube-system
  * myproject
    openshift
    openshift-infra

Using project "myproject".
[lmohanty@LalatenduM-laptop openshift-origin-client-tools-v1.5.0-rc.0-49a4a7a-linux-64bit]$ ./oc get svc docker-registry -n default
NAME              CLUSTER-IP   EXTERNAL-IP   PORT(S)    AGE
docker-registry   172.30.1.1   <none>        5000/TCP   1m

[lmohanty@LalatenduM-laptop openshift-origin-client-tools-v1.5.0-rc.0-49a4a7a-linux-64bit]$ ./oc login -u developer
Logged into "https://192.168.2.5:8443" as "developer" using existing credentials.

You have one project on this server: "myproject"

Using project "myproject".

$ OPENSHIFT_TOKEN=$(./oc whoami -t)

$ docker login -u developer -p ${OPENSHIFT_TOKEN} 172.30.1.1:5000
Login Succeeded

$ docker pull busybox
Using default tag: latest
Trying to pull repository docker.io/library/busybox ... 
sha256:32f093055929dbc23dec4d03e09dfe971f5973a9ca5cf059cbfb644c206aa83f: Pulling from docker.io/library/busybox
7520415ce762: Pull complete 
Digest: sha256:32f093055929dbc23dec4d03e09dfe971f5973a9ca5cf059cbfb644c206aa83f
Status: Downloaded newer image for docker.io/busybox:latest

$ docker tag busybox 172.30.1.1:5000/myproject/nginx
[lmohanty@LalatenduM-laptop openshift-origin-client-tools-v1.5.0-rc.0-49a4a7a-linux-64bit]$ docker push 172.30.1.1:5000/myproject/nginx
The push refers to a repository [172.30.1.1:5000/myproject/nginx]
c0de73ac9968: Pushed 
latest: digest: sha256:b52eee913b03dfc316e47b2eb05d33a79b6a018c85f2d30bdf3896cd6e22d408 size: 2098

```

```
 ./oc version
oc v1.5.0-rc.0+49a4a7a
kubernetes v1.5.2+43a9be4
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://192.168.2.5:8443
openshift v1.5.0-rc.0+49a4a7a
kubernetes v1.5.2+43a9be4
```

[1] https://github.com/openshift/origin/blob/master/docs/cluster_up_down.md#accessing-the-openshift-registry-directly

**Comment by praveenkumar** [comment]
What I think is we should secure our registry as steps provided by openshift documents since we used same steps during our ADB development cycle and it eliminate the putting insecure registry configuration to a VM host daemon and also we don't face issue which @LalatenduMohanty faced recently.

https://docs.openshift.org/latest/install_config/registry/securing_and_exposing_registry.html and https://github.com/projectatomic/adb-utils/blob/master/services/openshift/scripts/openshift_provision#L68-L116

**Comment by hferentschik** [comment]
@LalatenduMohanty - fyi, @praveenkumar and I made quite some progress on this issue as part of issue #177. See also pull request #601. Let's sync. There are still a couple of questions, but in general it should be possible to expose the secured registry.

**Comment by LalatenduMohanty** [comment]
With Origin 1.5 rc0 the unsecured registry is already exposed and we do not need to do any changes in Minishift for it. To use the exposed registry, user need to login to the registry. I think we should just add the piece of information as part of documentation.

Whether we need a secured docker registry for Minishift, I am not sure about it at this point. So we should find out from  Minishift users and OpenShift folks if we need secured registry in Minishift. @jorgemoralespou comments?

```
$ eval $(minishift docker-env)
$ OPENSHIFT_TOKEN=$(oc whoami -t)
$ docker login -u developer -p ${OPENSHIFT_TOKEN} 172.30.1.1:5000 
```

Tested docker push

```
$ docker pull busybox
$ docker tag busybox 172.30.1.1:5000/myproject/busybox
$ docker push 172.30.1.1:5000/myproject/busybox
```

**Comment by hferentschik** [comment]
So what's about we start with this and add appropriate documentation.

**Comment by LalatenduMohanty** [comment]
I am planning to send the doc PR after fix for #629 is merged.

**Comment by jorgemoralespou** [comment]
@LalatenduMohanty I don't see the value for a secure registry for local development environment. But it could be secure if we could inject our own certificates. 

Maybe I would add this to the roadmap, as a reminder, and see how things progress.

**Comment by hferentschik** [comment]
>  I don't see the value for a secure registry for local development environment. 

It is not so much about securing, but exposing a proper route. Using routes for non secured registries is not so easy, since you also need to re-configure the Docker daemon and theoretically we have all the right commands (see addon) to create and secure the registry route. It's just there still seems to be a puzzle bit missing.

Using a route is the proper dynamic way to get access to the registry. For example tooling like DevStudio could run to dynamically discover the route:

    $ minishift openshift service docker-registry --url

Right now, the best option for them is probably to hard code the IP and there potential problems can propagate. What happens when _cluster up_ changes from some reason the IP. Suddenly several components are broken.

So it's imo not so much about exposing a secured registry, but exposing the register the proper way, via a route and not rely on some _cluster up_ magic. We should abstract from this fact.

**Comment by LalatenduMohanty** [comment]
> So it's imo not so much about exposing a secured registry, but exposing the register the proper way, via a route and not rely on some cluster up magic. We should abstract from this fact.

+1. It might be worthy to check @csrwng why `oc cluster up` did not do it or they have plan to expose the route in future.

**Comment by csrwng** [comment]
@LalatenduMohanty there is no technical reason for the registry not being exposed.  At the time it wasn't a priority, but it should not be hard to do so.

**Comment by LalatenduMohanty** [comment]
After discussion with @csrwng we found that edge termination routes would be perfect fit for our use case to expose the docker-registry route. However unfortunately we there is an bug with the cluster up code [1] and which prevents us for doing this in Minishift. 

Steps for edge termination route for docker-registry in Minishift
 * Create the route
   $ oc create route edge docker-registry --service=docker-registry -n default
* Find the route
  $ oc get route docker-registry -o jsonpath='{.spec.host}' -n default
* Login to the Minishift VM and create the directory structure e.g. etc/docker/certs.d/docker-registry-default.\<IP\>.nip.io/
* Copy the master ca.crt (from /var/lib/minishift/openshift.local.config/master) to /etc/docker/certs.d/docker-registry-default.\<IP\>.nip.io/
* Log out of the VM and try to login to the docker-registry as developer (user)
   $ docker login -u developer -p oc whoami -t docker-registry-default..nip.io
 
[1] https://github.com/openshift/origin/issues/13646

**Comment by LalatenduMohanty** [comment]
Also `oc cluster up` planning to have route created for the docker-registry in future. Refer https://github.com/openshift/origin/issues/13643

**Comment by LalatenduMohanty** [comment]
FYI. @csrwng provided another was way to create the edge termination route https://github.com/openshift/origin/issues/13646#issuecomment-292228449

**Comment by hferentschik** [comment]
@praveenkumar, do I understand this correctly, that we have a solution now for doing edge terminated routes and exposing the registry as such a route?  Take it this happens via a modified version of the expose-registry plugin? 

What's about ISO compatibility? Does this work with Boot2Docker as well as CentOS ISO? Does it require a Docker daemon restart? Is it persistent?

**Comment by LalatenduMohanty** [comment]
One thing we should find out is if this works with OpenShift version 1.4.1. If yes, we can discuss if we want to also support 1.4.1. Refer https://github.com/minishift/minishift/issues/667

**Comment by praveenkumar** [comment]
> What's about ISO compatibility? Does this work with Boot2Docker as well as CentOS ISO? 

Yes checked with b2d and centos worked for both.

> Does it require a Docker daemon restart? 

No,

> Is it persistent

Yes,

@hferentschik One issue is if we have an addon enabled and restart minishift (start => stop => start) then addon again try to apply and with our case it fail because registry route already exist. Do we have any requirement why we need to reapply the addons since oc cluster up also don't do any provision when restart happen?

**Comment by LalatenduMohanty** [comment]
> we have an addon enabled and restart minishift (start => stop => start) then addon again try to apply and with our case it fail because registry route already exist. Do we have any requirement why we need to reapply

 That sounds like an issue. I think addon DSL should support this use case somehow i.e. conditional execution of statements.

**Comment by hferentschik** [comment]
> . I think addon DSL should support this use case somehow i.e. conditional execution of statements.

-1 No conditionals. There is something else going on.

**Comment by hferentschik** [comment]
Docs merged via #679
